plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/wizard/builder/ProjectBuilderTypeDetector.java

```diff
@@ -0,0 +1,107 @@
+ * and is available at http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package org.jboss.tools.openshift.internal.ui.wizard.builder;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+
+import org.eclipse.core.resources.IProject;
+import org.jboss.tools.openshift.common.core.utils.ProjectUtils;
+
+/**
+ * Detects the type of an {@link IProject}
+ * 
+ * @author Fred Bricon
+ */
+public class ProjectBuilderTypeDetector {
+
+	private static IProjectBuilderType UNKNOWN = new IProjectBuilderType(){
+		public boolean applies(IProject project) {
+			return true;
+		}
+
+		@Override
+		public String getTags(IProject project) {
+			return "";
+		};
+	};
+	private Collection<IProjectBuilderType> detectors;
+
+	public ProjectBuilderTypeDetector() {
+		//XXX might consider using Ext. Points in the future
+		//See Detection table in https://docs.openshift.org/latest/dev_guide/new_app.html#language-detection 
+		this(new SimpleTypeDetector("eap", "pom.xml"),//TODO be more specific wrt Tomcat 7/8, Micro services
+			 new SimpleTypeDetector("php", "index.php", "composer.json"),
+			 new SimpleTypeDetector("ruby", "Rakefile", "Gemfile", "config.ru"),
+			 new SimpleTypeDetector("python", "requirements.txt", "config.py"),
+			 new SimpleTypeDetector("perl", "index.pl", "cpanfile"),
+			 new SimpleTypeDetector("node", "app.json", "package.json")
+			 );
+	}
+
+	public ProjectBuilderTypeDetector(IProjectBuilderType...detectors) {
+		this.detectors = Arrays.asList(detectors);
+	}
+
+	protected IProjectBuilderType identify(final IProject project) {
+		if (project == null) {
+			return UNKNOWN;
+		}
+		return detectors.stream()
+						.filter(d -> d.applies(project))
+						.findFirst()
+						.orElse(UNKNOWN);
+	}
+
+	public String findTemplateFilter(final IProject project) {
+		if (project == null) {
+			return null;
+		}
+		return identify(project)
+				.getTags(project);//that's a bit ugly but we'll prolly need to get specific tags 
+		                          //depending on some other project settings
+	}
+
+	private static class SimpleTypeDetector implements IProjectBuilderType {
+
+		private String type;
+		private Collection<String> files;
+
+		SimpleTypeDetector(String type, String...files) {
+			this.type = type;
+			this.files = files == null ? Collections.emptyList():Arrays.asList(files);
+		}
+
+		@Override
+		public String getTags(IProject project) {
+			return type;
+		}
+
+		protected boolean hasAnyFile(final IProject project, Collection<String> files) {
+			return files.stream()
+						.filter(f -> project.getFile(f).exists())
+						.findFirst()
+						.isPresent();
+		}
+
+		@Override
+		public boolean applies(IProject project) {
+			return ProjectUtils.isAccessible(project) && hasAnyFile(project, files);
```

Comment:
what I grok here is that we rely on the presence of specific files to dertermine the type a project (php, node, etc.) rather than relying on a nature that must have been set by the user or automatically upon import. Shouldnt we use this as a fallback if there isnt a nature since the nature framework has more elaborate detection shemes (I guess).